### PR TITLE
Refactor Wagtail userbar as a web component

### LINKED ISSUE #9342: Wagtail userbar toggle doesn't close the userbar
### Issue Summary
When the user bar toggle button is toggled, the the items display but when it's toggled again and the items is supposed to disappear, they don't
 So when the user clicks on the user bar toggle button, the content displays fine. But when it's clicked again and the content is supposed to disappear, they don't. The content  disappears for some seconds and then reappears again. Then also, the first list item of has a yellow border around it when the toggle event is triggered.

### Steps to Reproduce

1. Start the wagtail bakery
2. Navigate to the fixed user bar nav on the bottom right corner
3. Toggle the button

Any other relevant information. For example, why do you consider this a bug and what did you expect to happen instead?

-   I expected that the second I click the button, the content is supposed to disappear or display none.

-   I have confirmed that this issue can be reproduced as described on a fresh Wagtail project: (yes / no) yes

user bar image 2

@thibaudcolas  @lb-  Please confirm. Thanks!

### PR DESCRIPTION
**Pull request by albinazs**
**Refactor Wagtail userbar - web component:**
- defines wagtail-userbar custom element
- initialises with open shadow DOM
- removes template from html after being used

**Refactor Wagtail userbar stylesheets to make use of web components isolation:**
- changes namespace to w-
- uses Wagtail’s main color variables, changing Tailwind’s :root variables assignment to :root,:host for Web Components compatibility
- removes -userbar-reset styles (and all other all declarations)
- removes most of !important from declarations
- switches font-size declarations to rem
- improves Windows High Contrast Mode design: bugfix for ::after element and added border

### CODE REVIEW

File: wagtail/admin/templates/wagtailadmin/userbar/base.html [hunk 1]
```diff
@@ -1,10 +1,10 @@
 {% load wagtailadmin_tags i18n %}
 <!-- Wagtail user bar embed code -->
-<div class="wagtail-userbar-reset">
-    <div class="wagtail-userbar wagtail-userbar--{{ position|default:'bottom-right' }}" data-wagtail-userbar>
+<template id="wagtail-userbar-template" class="w-userbar-reset">
```
Comment (thibaudcolas) [inline wagtail/admin/templates/wagtailadmin/userbar/base.html:3]: Can we remove `w-userbar-reset`?

File: client/src/entrypoints/admin/userbar.js [hunk 2]
```diff
@@ -5,249 +5,270 @@
 // This component implements a roving tab index for keyboard navigation
 // Learn more about roving tabIndex: https://w3c.github.io/aria-practices/#kbd_roving_tabindex
```
Comment (thibaudcolas) [inline client/src/entrypoints/admin/userbar.js:6]: Not for this specific line – I’m finding the JS really hard to review, I’m surprised there are so many changes. In the future I’d have really preferred if we kept the toggle fix and the refactoring completely separate, as:

- It’d have made it easier to review each change separately.
- It’d have made it possible for us to apply the bug fix to both Wagtail’s upcoming version and as a patch to the current release.

Generally, combining multiple unrelated changes just makes it harder for us to get them merged.

I’ll hopefully be able to come back to this later today and make sense of the changes as-is.

Comment (albinazs) [reply client/src/entrypoints/admin/userbar.js:6]: I committed all the suggestions except for the md file for writing templates - I need a little bit more time to test it.
I also re-opened my PR #9808 for toggle issue bugfix to separate unrelated changes.

Comment (albinazs) [reply client/src/entrypoints/admin/userbar.js:6]: I tested styling outside the web component, and came up with the solution in my last commit

File: client/src/entrypoints/admin/userbar.js [hunk 3]
```diff
@@ -5,249 +5,263 @@
...
-  function isFocusOnItems() {
-    return (
-      document.activeElement &&
-      !!document.activeElement.closest('.wagtail-userbar-items')
-    );
-  }
+    function isFocusOnItems() {
+      return (
+        shadowRoot.activeElement &&
+        !!shadowRoot.activeElement.closest('.w-userbar-items')
```
Comment (thibaudcolas) [inline client/src/entrypoints/admin/userbar.js:101]: Note will need to be updated to reflect the merged bug fix.

File: client/src/entrypoints/admin/userbar.js [hunk 4]
```diff
@@ -5,249 +5,263 @@
...
-    input:not([disabled])`;
+class Userbar extends HTMLElement {
+  connectedCallback() {
+    const template = document.getElementById('wagtail-userbar-template');
+    const shadowRoot = this.attachShadow({
+      mode: 'open',
+    });
+    shadowRoot.appendChild(template.content.cloneNode(true));
+    // Removes the template from html after it's being used
+    template.remove();
```
Comment (thibaudcolas) [inline client/src/entrypoints/admin/userbar.js:16]: Interesting, I wasn’t expecting this but it does feel much neater.

File: wagtail/admin/static_src/wagtailadmin/scss/userbar.scss [hunk 5]
```diff
@@ -279,9 +266,17 @@ $positions: (
...
     fill: $system-color-link-text;
   }
+
+  .w-userbar__item {
+    border: 1px solid $system-color-button-text;
+  }
+
+  .w-userbar-items::after {
+    border: $width-arrow solid Canvas;
+  }
```
Comment (thibaudcolas) [inline wagtail/admin/static_src/wagtailadmin/scss/userbar.scss:279]: Looking great!

**Review by thibaudcolas — APPROVED** [review]

**Pull request merged**